linux/src/completion_port_linux.c

```diff
@@ -0,0 +1,434 @@
+        }
+        else
+        {
+            // Codes_COMPLETION_PORT_LINUX_11_002: [ completion_port_create shall create the epoll instance by calling epoll_create. ]
+            result->epoll = epoll_create(MAX_EVENTS_NUM);
+            if (result->epoll == -1)
+            {
+                LogErrorNo("failure epoll_create MAX_EVENTS_NUM: %d", MAX_EVENTS_NUM);
+            }
+            else
+            {
+                // Codes_COMPLETION_PORT_LINUX_11_003: [ completion_port_create shall create a thread that runs epoll_worker_func to handle the epoll events. ]
+                if (ThreadAPI_Create(&result->thread_handle, epoll_worker_func, result) != THREADAPI_OK)
+                {
+                    LogCritical("Failure creating thread");
+                }
+                else
+                {
+                    // Codes_COMPLETION_PORT_LINUX_11_004: [ On success completion_port_create shall return the allocated COMPLETION_PORT_HANDLE. ]
+                    goto all_ok;
+                }
+            }
+            (void)close(result->epoll);
+        }
+        // Codes_COMPLETION_PORT_LINUX_11_005: [ If there are any errors then completion_port_create shall fail and return NULL. ]
+        REFCOUNT_TYPE_DESTROY(COMPLETION_PORT, result);
+        result = NULL;
+    }
+all_ok:
+    return result;
+}
+
+void completion_port_inc_ref(COMPLETION_PORT_HANDLE completion_port)
+{
+    // Codes_COMPLETION_PORT_LINUX_11_006: [ If completion_port is NULL, completion_port_inc_ref shall return. ]
+    if (completion_port == NULL)
+    {
+        LogError("Invalid arguments: COMPLETION_PORT_HANDLE completion_port=%p", completion_port);
+    }
+    else
+    {
+        // Codes_COMPLETION_PORT_LINUX_11_007: [ Otherwise completion_port_inc_ref shall increment the internally maintained reference count. ]
+        INC_REF(COMPLETION_PORT, completion_port);
+    }
+}
+
+void completion_port_dec_ref(COMPLETION_PORT_HANDLE completion_port)
+{
+    // Codes_COMPLETION_PORT_LINUX_11_008: [ If completion_port is NULL, completion_port_dec_ref shall return. ]
+    if (completion_port == NULL)
+    {
+        LogError("Invalid arguments: COMPLETION_PORT_HANDLE completion_port=%p", completion_port);
+    }
+    else
+    {
+        // Codes_COMPLETION_PORT_LINUX_11_009: [ completion_port_dec_ref shall decrement the reference count for completion_port. ]
+        if (DEC_REF(COMPLETION_PORT, completion_port) == 0)
+        {
+            // Codes_COMPLETION_PORT_LINUX_11_010: [ If the reference count reaches 0, completion_port_dec_ref shall do the following: ]
+            int32_t value;
+            // Codes_COMPLETION_PORT_LINUX_11_011: [ - wait for the ongoing call count to reach zero. ]
+            while ((value = interlocked_add(&completion_port->pending_calls, 0)) != 0)
+            {
+                (void)wait_on_address(&completion_port->pending_calls, value, UINT32_MAX);
+            }
```

Comment:
from what I see this is incremented and decremented in `completion_port_add`.

But there is no state change that happens before this check (like going to a "CLOSING" state).

Which means that someone could call `completion_port_add` after line 264 and it would work.

But this is a dec_ref so basically destroy, which means we are making this check during destroy.

As Andrei would say, destroy must be single threaded so no other thread should be accessing this object while it is being destroyed, so it seems like we should never execute line 263 (wait_on_address) because we can never have pending calls.

I think this check makes sense if this module has open/close. And if it does then it needs to do what we always did before sm - first set a state to close then wait for all calls to complete (and each call needs to first increment the pending count then check the state).

But as it is, I don't think it makes sense to have this pending call count at all. Otherwise I'm curious to see the test that would actually hit this `wait_on_address` line without artificially mocking `interlocked_add`.

Reply:
I need to move the worker_thread_continue increment up before line 260.  This would ensure that the add and remove won't race.

If I had two threads, one calling add and another calling dec_ref then you would need the pending call count (Dan brought this one up, so I can't take credit).  I will rearrange this a little submit.

Reply:
I guess my question is, how do you have this?

Reference count =1

T1: completion_port_add
T2: completion_port_dec_ref

I would argue that either T2 cannot call dec ref because it knows about the add on T1 (it holds the 1 and only reference afterall)

OR

T1 cannot call add because it should have had an inc_ref at some time (ref count should have been 2 so T2 would just decrement).

In other words, you can't call dec_ref until you know that you are no longer using (your reference to) the object and conversely you can't call anything on the object unless you know you have a reference.